### ISSUE TITLE
package.json with `"type": "module"` cause error when `amplify push`

### ISSUE THREAD

**Issue by yaquawa** [aws-amplify/amplify-cli#5691]
**Describe the bug**
package.json with `"type": "module"` be set cause the following error when I run `amplify push`

```
Must use import to load ES Module: /Users/x/Document/www/src/aws-exports.js
require() of ES modules is not supported.
require() of /Users/x/Document/www/src/aws-exports.js from /Users/x/.npm-global/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename aws-exports.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/x/Documents/www/package.json.

An error occurred during the push operation: Must use import to load ES Module: /Users/Document/www/aws-exports.js
require() of ES modules is not supported.
require() of /Users/x/Document/www/aws-exports.js from /Users/x/.npm-global/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename aws-exports.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/x/Documents/www/package.json.
```

**Amplify CLI Version**
4.30.0

**node.js Version**
v14.11.0

**Comment by jhockett** [comment]
Hi @yaquawa, can you please share more details about how this error came about? What does `amplify status` show?

**Comment by yaquawa** [comment]
@jhockett I think this is a webpack 5 compatibility problem.
I'm using [laravel-mix](https://github.com/JeffreyWay/laravel-mix) to bundle the js files, which it's utilizing webpack 5.

**Comment by jhockett** [comment]
I'm going to close this then @yaquawa. If it turns out that webpack 5 isn't causing the issue, comment below and the issue can be reopened.

**Comment by yaquawa** [comment]
Hi @jhockett I don't think this is an "issue" of WebPack, some library like [graphql-js](https://github.com/graphql/graphql-js) already supports webpack 5 as of its latest version.

I believe this can be fixed by simply changing `aws-exports.js` to `aws-exports.mjs` if the upstream(`amplify-cli`) is using commonjs.

By the way `amplify-cli` still depends on the old version of [graphql-js](https://github.com/graphql/graphql-js) so I have to replace the old one with the latest one manually.

**Comment by rodgco** [comment]
Hi @jhockett just to reinforce that this topic still not solved, and I agree with @yaquawa that this is not related to WebPack. I'm having the same issue trying to use the newly release SvelteKit with static distribution and console deploy.

It seems that as part of the "amplify push" process the aws-exports.js is imported by "amplify-frontend-javascript/lib/frontend-config-creator.js", and as the whole project is of type module a error is raised.

I took a look at the code and I found this piece where apparently there is a tentative to circumvent it, but it is not working. Couldn't deep dive on it to come with a PR.

```
    if (isPackaged) {
      // if packaged, we can't load an ES6 module because pkg doesn't support it yet
      const es5export = 'module.exports = {default: awsmobile};\n';
      const es6export = 'export default awsmobile;\n';
      const fileContents = fs.readFileSync(targetFilePath, 'utf-8');
      fs.writeFileSync(targetFilePath, fileContents.replace(es6export, es5export));
      awsExports = require(targetFilePath).default;
      fs.writeFileSync(targetFilePath, fileContents);
    } else {
      awsExports = require(targetFilePath).default;
    }
```

**Comment by jhockett** [comment]
Hi @rodgco, thanks for the detailed information. I'll add this to our triaging project so this can get better visibility.

Sorry @yaquawa, I missed your comment somehow and never followed up.

**Comment by cjihrig** [comment]
I don't think this is a Webpack issue. The [error message comes from Node](https://github.com/nodejs/node/blob/31fe3b215f175176d6474439f932bb5e7a12b573/lib/internal/errors.js#L1372-L1391). I think the long term solution is to move to `import()`, but I don't think we could get away with that yet due to varying ESM support in Node versions.

We might be able to do something similar to the code in https://github.com/aws-amplify/amplify-cli/issues/5691#issuecomment-808574722... read the file, write it back out as CJS to a file with the `.cjs` extension, require it, and then delete the `.cjs` file.

**Comment by asmajlovicmars** [comment]
Same error for me trying to push sveltekit static build with amplify.

``` 
Must use import to load ES Module: /Users/a/code/sveltekit/svk5/src/aws-exports.js
An error occurred during the push operation: Must use import to load ES Module: /Users/a/code/sveltekit/svk5/src/aws-exports.js
```

**Comment by EdwinFdoGarcia** [comment]
We are very excited with AWS Amplify
We started some academic exercises to deploy applications with Svelte-Kit in aws amplify, playing with Appi, storage ... etc.

Installation method: `mkdir my-app cd my-app npm init svelte @ next npm install npm run dev - --open`

- when starting the cloud environment it shows an error: (Error [ERR_REQUIRE_ESM])

<img width="1920" alt="112540154-d6db6e80-8d7f-11eb-88e2-cd6804dc5531" src="https://user-images.githubusercontent.com/35706123/113911169-43526680-979f-11eb-9bf5-b8177d75873a.png">
<img width="1912" alt="112533069-8f50e480-8d77-11eb-826b-8ea5fa1cab28" src="https://user-images.githubusercontent.com/35706123/113911174-464d5700-979f-11eb-9514-d47361180abd.png">

- We understand that it could be an Amplify ESM support error so we created the same academic exercise with Svelte-Sapper but now it works! We can deploy our academic project with Svelte-Sapper!

<img width="1918" alt="112533701-3cc3f800-8d78-11eb-96d0-8a538a949230" src="https://user-images.githubusercontent.com/35706123/113911397-8ca2b600-979f-11eb-8156-83e627ba5434.png">

**Comment by rodgco** [comment]
As far as I understand Sapper is not ESM based, so this test deviates from the original problem that is Amplify Support for ESM based projects.

**Comment by sinedie** [comment]
Hey there. Im using svelte kit with amplify. This error happens to me when I have already created `aws-exports.js`. When I delete it everything goes fine. I can do it manually when doing a change to the backend locally, but when using the hosting CI/CD I can not. Is there a workaround for this?, I just need to update my frontend, should be a simple node app (that build to static) and dont even need to update a single service on my backend. I can handle that by my own, but even if i dont put anything about the backend in the pipeline definition it tries to update the it or at least pull it.

**Comment by laverdet** [comment]
The problem here is that Amplify CLI writes out an ES6 module: [aws_exports.js.ejs](https://github.com/aws-amplify/amplify-cli/blob/eb9257eaee117d0ed53ebc23aa28ecd7b7510fa1/packages/amplify-frontend-javascript/lib/aws_exports.js.ejs) but then attempts to require it as a CommonJS module: [frontend-config-creator.js](https://github.com/aws-amplify/amplify-cli/blob/6334ea3a1941222711de3923dd0aa66ab50a5a64/packages/amplify-frontend-javascript/lib/frontend-config-creator.js#L226). It's like trying to execute a C++ program with nodejs, these are different file formats. There's documentation here: https://nodejs.org/api/packages.html

My suggestion to the Amplify team is to convert that template file into a CommonJS file and output with a `.cjs` extension. For the users of Amplify CLI I suggest replacing `export default awsmobile;` with `module.exports = awsmobile;` perhaps with an Amplify hook.

**Comment by asmajlovicmars** [comment]
Just tested this again, and it's interesting that updating aws-exports.js at the end of push works fine on an empty src project folder, but if sveltekit is in there it still gives an error. 

I thought it's the node version (v16.13.1) that was causing an issue, since I don't have a single require on this new project... 

Error: 

✖ There was an error initializing your environment.
require() of ES Module /Users/asmajlovic/code/adnan/blog/src/aws-exports.js not supported.
Instead change the require of aws-exports.js in null to a dynamic import() which is available in all CommonJS modules.
/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:1
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/asmajlovic/code/adnan/blog/src/aws-exports.js not supported.
Instead change the require of aws-exports.js in null to a dynamic import() which is available in all CommonJS modules.
    at getCurrentAWSExports (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:226:20)
    at getAWSExports (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:87:35)
    at createAWSExports (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:79:31)
    at Object.createFrontendConfigs (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/node_modules/amplify-frontend-javascript/index.js:47:16)
    at AmplifyToolkit.onCategoryOutputsChange [as _onCategoryOutputsChange] (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/extensions/amplify-helpers/on-category-outputs-change.js:44:37)
    at initializeEnv (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/initialize-env.js:80:31)
    at async pullBackend (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/pull-backend.js:30:5)
    at async Object.run (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/commands/pull.js:60:13)
    at async Object.executeAmplifyCommand (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/index.js:288:9)
    at async executePluginModuleCommand (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/execution-manager.js:164:5)
    at async executeCommand (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/execution-manager.js:35:9)
    at async Object.run (/Users/asmajlovic/.nvm/versions/node/v16.13.1/lib/node_modules/@aws-amplify/cli/lib/index.js:162:9)

**Comment by Nxtra** [comment]
> The problem here is that Amplify CLI writes out an ES6 module: [aws_exports.js.ejs](https://github.com/aws-amplify/amplify-cli/blob/eb9257eaee117d0ed53ebc23aa28ecd7b7510fa1/packages/amplify-frontend-javascript/lib/aws_exports.js.ejs) but then attempts to require it as a CommonJS module: [frontend-config-creator.js](https://github.com/aws-amplify/amplify-cli/blob/6334ea3a1941222711de3923dd0aa66ab50a5a64/packages/amplify-frontend-javascript/lib/frontend-config-creator.js#L226). It's like trying to execute a C++ program with nodejs, these are different file formats. There's documentation here: https://nodejs.org/api/packages.html
> 
> My suggestion to the Amplify team is to convert that template file into a CommonJS file and output with a `.cjs` extension. For the users of Amplify CLI I suggest replacing `export default awsmobile;` with `module.exports = awsmobile;` perhaps with an Amplify hook.

This hook solutions sounds reasonable. Changing manually all the time would be a headache.
Which hooks are we talking about here? `post-push.js` and `post-pull.js` maybe?

**Comment by jacksonkernion** [comment]
Any updates on this? I've tried a bunch of different suggestions in this thread but my SvelteKit app will not build properly on Amplify.

**Comment by jetslambda** [comment]
What worked for me was changing package.json type: "module" to type: "commonjs" of course that changes babel and nodes interpretation of the code so it won't work for everyone. Leaving this as a suggestion that may help others.

**Comment by kriti-eco** [comment]
I'm trying to follow steps mentioned here - https://docs.amplify.aws/cli/function/#graphql-from-lambda to call GraphQL API from lambda.
As soon as I add "type": "module" to package.json , it starts throwing an error on amplify mock command. As per above link I need to import node-fetch but I'm not able to do it. Lambda code is just a hello-world code as of now, just changing package.json starts creating this problem. 

stack: 'Error: Could not load lambda handler function due to Error [ERR_REQUIRE_ESM]: Must use import to load ES Module

**Comment by laverdet** [comment]
@kriti-eco try installing node-fetch@2 instead. Version 3 made a strict cutover to modules. I recommend kicking the can down the road until Amplify can better support modules.

**Comment by hanna-becker** [comment]
None of the workarounds mentioned here work for our use case. We now have pre/post push/pull hooks to replace the type from module to commonjs before the operation and back to module afterwards to get the CI build working. There were several other places where buggy/non-existing ESM support in Amplify caused trouble after we upgraded to Angular 13/14 that only ships ES modules. I really hope proper ESM support is something the Amplify team will address, soon.

**Comment by laverdet** [comment]
> I really hope proper ESM support is something the Amplify team will address, soon.

I have bad news for you

**Comment by vadimmelnicuk** [comment]
Any updates on this?

**Comment by michaelcuneo** [comment]
@vadimmelnicuk I've managed to get this working from the following flow... 

1: amplify init or amplify init project then let it fail.
2: rename aws-exports to .cjs extension.
3: copy aws-exports to .js so there is both a .cjs and .js in the project.
4: Remove type: "module" from package.json.
5: All subsequent pushes will be successful.

**Comment by vadimmelnicuk** [comment]
@michaelcuneo Brilliant, thanks. It worked :)

**Comment by amreddys** [comment]
I was able to use hooks to rename .js to .cjs and it is working fine in my Vite project. I was even able to deploy it to amplify hosting but subsequent CD builds keep failing as amplify tries to initialize environment after pull but before post-pull hook runs.

As it works the first time I want to atleast delete the entire hosting and recreate for deployment but I couldnt find a delete for just the frontend.

Can anyone suggest some way to rename during build before the cli freaks out about the file?

# Starting phase: build
2022-08-16T12:41:18.656Z [INFO]: 
2022-08-16T12:41:18.657Z [INFO]: ----- 🪝 pre-pull execution start -----
2022-08-16T12:41:18.705Z [INFO]: file doesnt exist at /codebuild/output/src337420161/src/project/dir
2022-08-16T12:41:18.709Z [INFO]: ----- 🪝 pre-pull execution end -----
2022-08-16T12:41:18.710Z [INFO]: 
2022-08-16T12:41:19.558Z [INFO]: [0mAmplify AppID found: d3oasdzaluk345. Amplify App name is: project[0m
2022-08-16T12:41:19.602Z [INFO]: [0mBackend environment main found in Amplify Console app: project[0m
2022-08-16T12:41:20.480Z [WARNING]: - Fetching updates to backend environment: main from the cloud.
2022-08-16T12:41:21.198Z [WARNING]: - Building resource auth/userPoolGroups
2022-08-16T12:41:21.347Z [WARNING]: - Building resource auth/projectCognito
2022-08-16T12:41:21.465Z [WARNING]: ✔ Successfully pulled backend environment main from the cloud.
2022-08-16T12:41:21.483Z [WARNING]: ✖ There was an error initializing your environment.
2022-08-16T12:41:21.501Z [INFO]: [31mFailed to pull the backend.[39m
2022-08-16T12:41:21.505Z [INFO]: 🛑 Must use import to load ES Module: /codebuild/output/src337420161/src/project/dir/aws-exports.js
                                 require() of ES modules is not supported.
                                 require() of /codebuild/output/src337420161/src/project/dir/aws-exports.js from /snapshot/repo/build/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
                                 Instead rename aws-exports.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /codebuild/output/src337420161/src/project/dir/package.json.
2022-08-16T12:41:21.505Z [INFO]: Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /codebuild/output/src337420161/src/project/dir/aws-exports.js
                                 require() of ES modules is not supported.
                                 require() of /codebuild/output/src337420161/src/project/dir/aws-exports.js from /snapshot/repo/build/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
                                 Instead rename aws-exports.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /codebuild/output/src337420161/src/project/dir/package.json.
                                 at new NodeError (internal/errors.js:322:7)
                                 at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
                                 at Module.load (internal/modules/cjs/loader.js:950:32)
                                 at Function.Module._load (internal/modules/cjs/loader.js:790:12)
                                 at Module.require (internal/modules/cjs/loader.js:974:19)
                                 at Module.require (pkg/prelude/bootstrap.js:1719:31)
                                 at require (internal/modules/cjs/helpers.js:93:18)
                                 at getCurrentAWSExports (/snapshot/repo/build/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:296:18)
                                 at getAWSExports (/snapshot/repo/build/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:153:35)
                                 at createAWSExports (/snapshot/repo/build/node_modules/amplify-frontend-javascript/lib/frontend-config-creator.js:145:31)
                                 at Object.createFrontendConfigs (/snapshot/repo/build/node_modules/amplify-frontend-javascript/index.js:47:16)
                                 at AmplifyToolkit.onCategoryOutputsChange [as _onCategoryOutputsChange] (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/extensions/amplify-helpers/on-category-outputs-change.js:44:37)
                                 at initializeEnv (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/initialize-env.js:83:31)
                                 at processTicksAndRejections (internal/process/task_queues.js:95:5)
                                 at async onSuccess (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/attach-backend.js:94:9)
                                 at async attachBackend (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/attach-backend.js:45:9)
                                 at async Object.run (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/commands/pull.js:70:9)
                                 at async Object.executeAmplifyCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/index.js:292:9)
                                 at async executePluginModuleCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/execution-manager.js:142:5)
                                 at async executeCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/execution-manager.js:40:9)
                                 at async Object.run (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/index.js:165:13)
2022-08-16T12:41:21.506Z [INFO]: 
2022-08-16T12:41:21.506Z [INFO]: ----- 🪝 post-pull execution start -----
2022-08-16T12:41:21.554Z [INFO]: file exists at /codebuild/output/src337420161/src/project/dir renaming
2022-08-16T12:41:21.556Z [INFO]: null
2022-08-16T12:41:21.563Z [INFO]: ----- 🪝 post-pull execution end -----
2022-08-16T12:41:21.563Z [INFO]: 
2022-08-16T12:41:21.581Z [ERROR]: !!! Build failed
2022-08-16T12:41:21.581Z [ERROR]: !!! Non-Zero Exit Code detected

**Comment by Karol-Perec** [comment]
Hello everyone, if you want to have ES Modules working with amplify in CI pipeline, you need to change `amplifyPush` script behaviour. I tweaked it in my own project. It is available in [https://github.com/Karol-Perec/wykopX/blob/master/scripts/amplifyPush.sh](url). So you just need to add this custom script to your repository, for example in `$root/scripts/amplifyPush.sh`, just like i did. Then you need to update your CI backend build pipeline to execute custom script instead of the built-in one: 
```
backend:
  phases:
    build:
      commands:
        - sh scripts/amplifyPush.sh --simple
```
TLDR: i just added renaming `aws-exports.js` to `aws-exports.cjs` before running `amplify init`, so every of your CI pipelines will work with ES Modules out of the box.

**Comment by bart** [comment]
Not sure why this issue got closed. I'm using `"type": "module"` in my package.json (not possible to change it because the entire app depends on that setting). When I push changes to my Github repo Amplify CI/CD deployment starts and fails with the following error: `Must use import to load ES Module`

It makes no sense to rename aws-exports.js to .cjs locally because during push or pull commands it's generating .js files locally instead. My build config is a default Vue 3 one:

```
version: 1
backend:
  phases:
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - amplifyPush --simple
frontend:
  phases:
    preBuild:
      commands:
        - npm install
    build:
      commands:
        - npm run build
  artifacts:
    baseDirectory: dist
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*
```

Any idea how to solve this @danielleadams @armenr ? Any help is highly appreciated, thanks!

**Comment by bart** [comment]
Thank you so much @armenr for sharing your hook scripts and diving deep into the topic. I already understood the hooks concept but after running into another Amplify issue (where I would have to find another workaround for) I decided to go the good old fashioned way using Serverless framework. In my opinion Amplify unfortunately is still not stable and flexible enough to get used in enterprise applications. But for sure this is just a personal opinion and I'm sure other devs would love to go this hook way. So again, thanks a lot for sharing!!

**Comment by bart** [comment]
Thanks a lot @armenr for pointing me into the SST / CDK direction. Accidentally I already tinkered around with it yesterday. Like the simplicity but decided to go with Serverless framework in the end because I'm most familiar with it. But maybe for a side project or smaller project in future SST might be a good choice.

**Comment by josefaidt** [comment]
Hey @bart and @armenr this fix has been released with the [latest version of Amplify CLI, 10.2.3](https://www.npmjs.com/package/@aws-amplify/cli) 🙂

**Comment by hanna-becker** [comment]
Really happy this got fixed!
